kvclient/kvcoord: inhibit parallel commit when retrying EndTxn request

andreimatei opened the pull request:
The scenario that this patch addresses is the following (from #46431):
1. txn1 sends Put(a) + Put(b) + EndTxn
2. DistSender splits the Put(a) from the rest.
3. Put(a) succeeds, but the rest catches some retriable error.
4. TxnCoordSender gets the retriable error. The fact that a sub-batch
  succeeded is lost. We used to care about that fact, but we've
  successively gotten rid of that tracking across #35140 and #44661.
5. we refresh everything that came before this batch. The refresh
  succeeds.
6. we re-send the batch. It gets split again. The part with the EndTxn
  executes first. The transaction is now STAGING. More than that, the txn
  is in fact implicitly committed - the intent on a is already there since
  the previous attempt and, because it's at a lower timestamp than the txn
  record, it counts as golden for the purposes of verifying the implicit
  commit condition.
7. some other transaction wonders in, sees that txn1 is in its way, and
  transitions it to explicitly committed.
8. the Put(a) now tries to evaluate. It gets really confused. I guess
  that different things can happen; none of them good. One thing that I
  believe we've observed in #46299 is that, if there's another txn's
  intent there already, the Put will try to push it, enter the
  txnWaitQueue, eventually observe that its own txn is committed and
  return an error. The client thus gets an error (and a non-ambiguous one
  to boot) although the txn is committed. Even worse perhaps, I think it's
  possible for a request to return wrong results instead of an error.

This patch fixes it by inhibiting the parallel commit when the EndTxn
batch is retried. This way, there's never a STAGING record.

Release note (bug fix): A rare bug causing errors to be returned for
successfully committed transactions was fixed. The most common error
message was "TransactionStatusError: already committed".

Release justification: serious bug fix

Fixes #46341

andreimatei committed:
kvclient/kvcoord: inhibit parallel commit when retrying EndTxn request

The scenario that this patch addresses is the following (from #46431):
1. txn1 sends Put(a) + Put(b) + EndTxn
2. DistSender splits the Put(a) from the rest.
3. Put(a) succeeds, but the rest catches some retriable error.
4. TxnCoordSender gets the retriable error. The fact that a sub-batch
  succeeded is lost. We used to care about that fact, but we've
  successively gotten rid of that tracking across #35140 and #44661.
5. we refresh everything that came before this batch. The refresh
  succeeds.
6. we re-send the batch. It gets split again. The part with the EndTxn
  executes first. The transaction is now STAGING. More than that, the txn
  is in fact implicitly committed - the intent on a is already there since
  the previous attempt and, because it's at a lower timestamp than the txn
  record, it counts as golden for the purposes of verifying the implicit
  commit condition.
7. some other transaction wonders in, sees that txn1 is in its way, and
  transitions it to explicitly committed.
8. the Put(a) now tries to evaluate. It gets really confused. I guess
  that different things can happen; none of them good. One thing that I
  believe we've observed in #46299 is that, if there's another txn's
  intent there already, the Put will try to push it, enter the
  txnWaitQueue, eventually observe that its own txn is committed and
  return an error. The client thus gets an error (and a non-ambiguous one
  to boot) although the txn is committed. Even worse perhaps, I think it's
  possible for a request to return wrong results instead of an error.

This patch fixes it by inhibiting the parallel commit when the EndTxn
batch is retried. This way, there's never a STAGING record.

Release note (bug fix): A rare bug causing errors to be returned for
successfully committed transactions was fixed. The most common error
message was "TransactionStatusError: already committed".

Release justification: serious bug fix

Fixes #46341

andreimatei committed:
util/tracing: move extractMsgFromRecord to the tracking pkg

This function was used in sql, but I want to start using it elsewhere
when dealing with traces.

Release note: None
Release justification: Minor code move, in support of the next commit.

andreimatei committed:
kvserver/batcheval: add a comment

Release note: None
Release justification: Comment only.

andreimatei committed:
kv, server: minor tracing improvements

Release note: None
Release justification: Minor new log messages.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46596)
<!-- Reviewable:end -->

andreimatei commented:
@nvanbenschoten was there any saving grace for not having to backport this to 19.2?

I still need to check if this actually fixes kv/contention (#45700).

nvanbenschoten commented:
Reviewing now.

> was there any saving grace for not having to backport this to 19.2?

Yes, we still had the MixedSuccessError back in 19.2, which is why we didn't hit this issue back then.

nvb reviewed (commented):
Nice, thanks for putting this together so fast. I feel good about how small this is, given where we are in the release cycle.

> Reviewed 1 of 1 files at r1, 3 of 3 files at r2, 4 of 4 files at r3, 3 of 3 files at r4, 3 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-:-M3MqlC6BG849Bb9K5Nu:b-z2qcx2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2616 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3Mn0FIDvfESFJw3Sni:-M3Mn0FJ3pWlyJRpxpgZ:b-s5ctqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7380c63d5471e4f514860b1d813a0f73b7049e/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2616)):*
> ```Go
> }
> 
> // Test that, even though at the kvserver level requests are not idempotent
> ```

Does this test fail without your change?
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2682 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MmbUh6CIB6F-QA27X:-M3MmbUh6CIB6F-QA27Y:b-mn9fc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7380c63d5471e4f514860b1d813a0f73b7049e/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2682)):*
> ```Go
> 	require.NoError(t, err)
> 
> 	// After the txn started, do a conflicting write. This will cause the txn's
> ```

"a conflicting read"
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2685 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MmfZn3OYZ7Y7AqZKa:-M3MmfZn3OYZ7Y7AqZKb:bwl8omw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7380c63d5471e4f514860b1d813a0f73b7049e/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2685)):*
> ```Go
> 	// upcoming left half to return a WriteTooOldError on the first attempt,
> 	// causing in turn a refresh and a retry. Note that the upcoming write is a
> 	// CPut, so it can't return the WriteTooOld flag instead of a
> ```

Use the word "defer" somewhere in here, since that's what we seem to be calling this.
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2582 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MnqXUBKukmon8QpKo:-M3MnqXV5kuYA46vSsko:boa6p5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2582)):*
> ```Go
> 	// pushee txn record.
> 	expectPusheeTxnRecordNotFound
> 	dontExpectNothing
> ```

Give this a comment.
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2658 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MnzOA7WWC--7HQNAa:-M3MnzOA7WWC--7HQNAb:b-xj2gc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2658)):*
> ```Go
> //    previous partial success).
> // 5. The batch is split again. b's sub-batch executed first. Say a's batch is
> // 	  delayed.
> ```

What's up with the strange spacing in these comments (and below in the next test)?
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2662 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3Mon0CDlgVfeF1zszZ:-M3Mon0CDlgVfeF1zsz_:b-rk9sb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2662)):*
> ```Go
> // The point of this test is to check that the txn is not considered to be
> // implicitly committed at this point. Handling this scenario requires special
> // care. If we wouldn't do anything, then we'd end up with a STAGING txn record
> ```

s/wouldn't/didn't/
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2674 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MpEUz7TkxwPfhjiOY:-M3MpEUz7TkxwPfhjiOZ:b-yrfbcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2674)):*
> ```Go
> // attempt. Thus, the EndTxn will be split from all the other requests, and the
> // txn record is never written if anything fails.
> func TestTxnCoordSenderRetriesAcrossEndTxn_EndTxnSucceedsLate(t *testing.T) {
> ```

I'm surprised you're not able to combine these two tests into a single test with two subtests. Don't they only differ in three places?
1. where the intermediate read is
2. which write is eventually rejected
3. whether we expect to find the txn record

This indicates that we should test all four combinations of the first two differences.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go, line 154 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MqIbSEZQMF_5xccMg:-M3MqIbT2LQDWCaM1IB0:b4cj3uj) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go#L154)):*
> ```Go
> 		// could technically enter the "implicitly committed" state before all the
> 		// sub-batches are evaluated and this is problematic: there's a race between
> 		// evaluating those requests randos coming along and transitioning the txn
> ```

"requests randos" is missing a word.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go, line 158 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MqNyu5zDGGDf24HPP:-M3MqNyu5zDGGDf24HPQ:b-8sl15f) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go#L158)):*
> ```Go
> 		// evaluations of the outstanding sub-batches. If the randos win, then the
> 		// re-evaluations will fail because we don't have idempotency of evaluations
> 		// across a txn commit.
> ```

Also mention that the re-evaluations might notice that their txn was committed and throw the error we see.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go, line 164 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3Mq_1t-7QzCbevRk1C:-M3Mq_1t-7QzCbevRk1D:b229m1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go#L164)):*
> ```Go
> 			// disable the parallel commit.
> 			etCpy := *et
> 			ba.Requests[len(ba.Requests)-1].SetInner(&etCpy)
> ```

Do we need to copy the entire `ba.Requests` slice to keep the race detector happy?
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go, line 301 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MprVI3zt-CM7SaG1b:-M3MprVI3zt-CM7SaG1c:b-hj1yl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go#L301)):*
> ```Go
> }
> 
> type endTxnAttempt int
> ```

We should either give this a comment or just replace it with a boolean. The enum seems somewhat overkill.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go, line 315 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MpngNDYTRlgnCZzpW:-M3MpngNDYTRlgnCZzpX:bt6l01n) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go#L315)):*
> ```Go
> 	ctx context.Context, ba roachpb.BatchRequest, et *roachpb.EndTxnRequest, etAttempt endTxnAttempt,
> ) bool {
> 	if etAttempt == endTxnRetry {
> ```

nit: move this down below `if !et.Commit {`. There's an order of precedence here.
___
*[pkg/kv/kvserver/batcheval/cmd_push_txn.go, line 143 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3Mjowz33ZqUnQhcMDj:-M3Mjowz33ZqUnQhcMDk:b378843) ([raw file](https://github.com/cockroachdb/cockroach/blob/a896174c0599004275c5258362ba258b64dc2ef3/pkg/kv/kvserver/batcheval/cmd_push_txn.go#L143)):*
> ```Go
> 	ok, err := storage.MVCCGetProto(
> 		ctx, readWriter, key,
> 		hlc.Timestamp{}, // the txn record is an "inline value", not MVCC
> ```

:thumbsdown: we do this all over the place, so adding a comment to this one place doesn't help. If you want to do something here, consider instead addin a comment on `MVCCGet` that discusses inline values like we have on `MVCCPut`.
___
*[pkg/kv/kvserver/batcheval/cmd_push_txn.go, line 266 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MlMUc9SCMPX36mchi:-M3MlMUdDXCufnZ1O3Vu:b-x47l47) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7380c63d5471e4f514860b1d813a0f73b7049e/pkg/kv/kvserver/batcheval/cmd_push_txn.go#L266)):*
> ```Go
> 	if reply.PusheeTxn.Status == roachpb.STAGING && (pusherWins || recoverOnFailedPush) {
> 		err := roachpb.NewIndeterminateCommitError(reply.PusheeTxn)
> 		log.VEventf(ctx, 1, "%v", err)
> ```

If you're going to do that here, do the same down below for the txn push error.
___
*[pkg/util/tracing/recorded_span.go, line 18 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3Ml9GAETwsLTpf2Gew:-M3Ml9GAETwsLTpf2Gex:b-uwlzl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/52f7f252ee219b4fbf6daa430459f6c658abdf26/pkg/util/tracing/recorded_span.go#L18)):*
> ```Go
> )
> 
> const LogMessageField string = "event"
> ```

No need for "string"
___
*[pkg/util/tracing/recorded_span.go, line 37 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MlAmZArzcUt7jWuuX:-M3MlAm_C-vsr3RP9tyj:bimw108) ([raw file](https://github.com/cockroachdb/cockroach/blob/52f7f252ee219b4fbf6daa430459f6c658abdf26/pkg/util/tracing/recorded_span.go#L37)):*
> ```Go
> 			return f.Value
> 		}
> 		if key == "error" {
> ```

Consider pulling "error" into a constant as well.
___
*[pkg/util/tracing/tracer_span.go, line 352 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3Mljzu0dCgmTNA0v8o:-M3Mljzu0dCgmTNA0v8p:b-jphbj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7380c63d5471e4f514860b1d813a0f73b7049e/pkg/util/tracing/tracer_span.go#L352)):*
> ```Go
> // FindLogMessage returns the first log message in the recording that matches
> // the given regexp. If no such messages exist, returns "".
> func (r Recording) FindLogMessage(pattern string) string {
> ```

Consider also returning a bool so we're not relying on string comparisons.


<!-- Sent from Reviewable.io -->

andreimatei committed:
kvclient/kvcoord: display traces differently

Before this patch, a trace (recording) was String()ed by sorting
log messages by timestamp, across spans. This proved to be a terrible
idea. This patch moves to keeping requests from the same span together,
but still interspersing children spans in the parent. So messages from
siblings are no longer mixed together.

Release note: None
Release justification: testing only

andreimatei committed:
kvcoord: add a log msg

Release note: None
Release justification: low risk, just logging

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46596)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @nvanbenschoten)

___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2616 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3Mn0FIDvfESFJw3Sni:-M3NP7gM5k2Xc9kAkXrX:b-1o7tse) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7380c63d5471e4f514860b1d813a0f73b7049e/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2616)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this test fail without your change?
</blockquote></details>

`TestTxnCoordSenderRetriesAcrossEndTxn_EndTxnSucceedsLate` fails, yes.
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2682 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MmbUh6CIB6F-QA27X:-M3lp6NACsoVbD_HNwKS:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7380c63d5471e4f514860b1d813a0f73b7049e/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2682)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"a conflicting read"
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2685 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MmfZn3OYZ7Y7AqZKa:-M3lpGj21SJVXVOboSXl:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7380c63d5471e4f514860b1d813a0f73b7049e/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2685)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Use the word "defer" somewhere in here, since that's what we seem to be calling this.
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2582 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MnqXUBKukmon8QpKo:-M3ly1du-gNqNkiZkLH9:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2582)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Give this a comment.
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2658 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MnzOA7WWC--7HQNAa:-M3lojk6BLkGhYuL0sBz:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2658)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What's up with the strange spacing in these comments (and below in the next test)?
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2662 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3Mon0CDlgVfeF1zszZ:-M3lorOj-eXh97lngSYp:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2662)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/wouldn't/didn't/
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2674 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MpEUz7TkxwPfhjiOY:-M3losAw39cQ7LchvHlL:bo4koi4) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2674)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm surprised you're not able to combine these two tests into a single test with two subtests. Don't they only differ in three places?
1. where the intermediate read is
2. which write is eventually rejected
3. whether we expect to find the txn record

This indicates that we should test all four combinations of the first two differences.
</blockquote></details>

done, except I'm only testing two of the 4 combinations. The others seem to trivial to me / they covered in more mundane tests.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go, line 154 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MqIbSEZQMF_5xccMg:-M3lpST5Cyr0nU2fcL03:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go#L154)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"requests randos" is missing a word.
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go, line 158 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MqNyu5zDGGDf24HPP:-M3lppN95oUZhxAafUa0:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go#L158)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Also mention that the re-evaluations might notice that their txn was committed and throw the error we see.
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go, line 164 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3Mq_1t-7QzCbevRk1C:-M3lrAE9-tEnv3VxKeCX:b-xh4lv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go#L164)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need to copy the entire `ba.Requests` slice to keep the race detector happy?
</blockquote></details>

I don't think so. The race transport claims that the client is already mutating the request slices of requests.
https://github.com/cockroachdb/cockroach/blob/b48fa5ae061366229d746cb91a43cd7326028a81/pkg/kv/kvclient/kvcoord/transport_race.go#L58
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go, line 301 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MprVI3zt-CM7SaG1b:-M3lrc1DA_7lpdLRynJL:b-4mhpgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go#L301)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We should either give this a comment or just replace it with a boolean. The enum seems somewhat overkill.
</blockquote></details>

added a comment
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go, line 315 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MpngNDYTRlgnCZzpW:-M3lxkzj3YtaKF8tjZoV:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/af92767df22bbb26131b07d78effb9ad31be14bb/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go#L315)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: move this down below `if !et.Commit {`. There's an order of precedence here.
</blockquote></details>

done
___
*[pkg/kv/kvserver/batcheval/cmd_push_txn.go, line 143 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3Mjowz33ZqUnQhcMDj:-M3NNmg_45NOo2LIUpu4:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a896174c0599004275c5258362ba258b64dc2ef3/pkg/kv/kvserver/batcheval/cmd_push_txn.go#L143)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

:thumbsdown: we do this all over the place, so adding a comment to this one place doesn't help. If you want to do something here, consider instead addin a comment on `MVCCGet` that discusses inline values like we have on `MVCCPut`.
</blockquote></details>

done
___
*[pkg/kv/kvserver/batcheval/cmd_push_txn.go, line 266 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MlMUc9SCMPX36mchi:-M3lyCWuEdLzFpKICTkP:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7380c63d5471e4f514860b1d813a0f73b7049e/pkg/kv/kvserver/batcheval/cmd_push_txn.go#L266)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If you're going to do that here, do the same down below for the txn push error.
</blockquote></details>

done
___
*[pkg/util/tracing/recorded_span.go, line 37 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MlAmZArzcUt7jWuuX:-M3NNzmR1SRxgYDrBjfN:b-s5f5f0) ([raw file](https://github.com/cockroachdb/cockroach/blob/52f7f252ee219b4fbf6daa430459f6c658abdf26/pkg/util/tracing/recorded_span.go#L37)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider pulling "error" into a constant as well.
</blockquote></details>

I couldn't tell how used this "error" key is, that's why I left it alone.
___
*[pkg/util/tracing/tracer_span.go, line 352 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3Mljzu0dCgmTNA0v8o:-M3lJ-vN2iMsUUJu3vXi:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7380c63d5471e4f514860b1d813a0f73b7049e/pkg/util/tracing/tracer_span.go#L352)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider also returning a bool so we're not relying on string comparisons.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46596)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/util/tracing/recorded_span.go, line 18 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3Ml9GAETwsLTpf2Gew:-M3lyxBvCYC5kP5ykvap:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/52f7f252ee219b4fbf6daa430459f6c658abdf26/pkg/util/tracing/recorded_span.go#L18)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

No need for "string"
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 11 of 11 files at r6, 3 of 3 files at r7, 4 of 4 files at r8, 1 of 1 files at r9, 1 of 1 files at r10, 4 of 4 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46596)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2648 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mLv0s241af0h4utHn:-M3mLv0tD--2x5Y2e7gR:b39fjne) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c8fc6fb2795e21be9c7f99a93cf5a65f961fc3d/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2648)):*
> ```Go
> )
> 
> type expectedTxnResolution int
> ```

Why do we need the enum? Can't we just use a `roachpb.TransactionStatus`?
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2749 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mNDil9CefP6ZhC0le:-M3mNDil9CefP6ZhC0lf:b230yna) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c8fc6fb2795e21be9c7f99a93cf5a65f961fc3d/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2749)):*
> ```Go
> // intent written in the 2nd attempt satisfying a STAGING record written on the
> // first attempt). See subtests for more details.
> func TestTxnCoordSenderRetriesAcrossEndTxn(t *testing.T) {
> ```

Nice, this test came out well.
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2771 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mMw2J5CuMOdoyY2jB:-M3mMw2J5CuMOdoyY2jC:b-n2kah5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c8fc6fb2795e21be9c7f99a93cf5a65f961fc3d/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2771)):*
> ```
> // The left side is CPut(a), the right side is CPut(b)+EndTxn(STAGING).
> ```

Consider moving this up to the `side` type definition.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go, line 211 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mLdlo4pTAfemr12VQ:-M3mLdlo4pTAfemr12VR:b-vmlm9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/00c32164199a475bbf96c2cf4f7cf0629c40064f/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go#L211)):*
> ```Go
> 	// DistSender.
> 	if pErr := needTxnRetryAfterStaging(br); pErr != nil {
> 		log.VEventf(ctx, 2, "parallel commit failed since some writes were pushed")
> ```

Maybe we should include the error in here.
___
*[pkg/util/tracing/recorded_span.go, line 37 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3MlAmZArzcUt7jWuuX:-M3mI2kM6SscOuEdUnUc:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/52f7f252ee219b4fbf6daa430459f6c658abdf26/pkg/util/tracing/recorded_span.go#L37)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I couldn't tell how used this "error" key is, that's why I left it alone.
</blockquote></details>

Ack.
___
*[pkg/util/tracing/tracer_span.go, line 264 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mIwWj4ymbPRy9JuKc:-M3mIwWj4ymbPRy9JuKd:boa6p5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/171b4e817ffa5e46f5b828242fdbd4cddd82f65f/pkg/util/tracing/tracer_span.go#L264)):*
> ```Go
> 	// log 1
> 	//   start span B  // duration relative to "log 1"
> 	//   log 2
> ```

Give this a comment.
___
*[pkg/util/tracing/tracer_span.go, line 290 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mKddm4NNkcW8Ri4n4:-M3mKddm4NNkcW8Ri4n5:b-tijdke) ([raw file](https://github.com/cockroachdb/cockroach/blob/171b4e817ffa5e46f5b828242fdbd4cddd82f65f/pkg/util/tracing/tracer_span.go#L290)):*
> ```Go
> // TODO(andrei): this should be unified with
> // SessionTracing.GenerateSessionTraceVTable.
> func (r Recording) String() string {
> ```

Can we write a test for this? Mostly I'm interested in seeing what effect it had.
___
*[pkg/util/tracing/tracer_span.go, line 302 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mJcNN8u68bTZELhn6:-M3mJcNN8u68bTZELhn7:b-fbp2hg) ([raw file](https://github.com/cockroachdb/cockroach/blob/171b4e817ffa5e46f5b828242fdbd4cddd82f65f/pkg/util/tracing/tracer_span.go#L302)):*
> ```Go
> 	}
> 
> 	var buf bytes.Buffer
> ```

`strings.Builder`
___
*[pkg/util/tracing/tracer_span.go, line 334 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mKMyc3XNS_rmSCOCs:-M3mKMyc3XNS_rmSCOCt:b-pnw391) ([raw file](https://github.com/cockroachdb/cockroach/blob/171b4e817ffa5e46f5b828242fdbd4cddd82f65f/pkg/util/tracing/tracer_span.go#L334)):*
> ```Go
> }
> 
> func (r Recording) visitSpan(sp RecordedSpan, depth int) []traceLogData {
> ```

Give this a comment. Also, to be honest, I'm having a really hard time understanding what changed in this commit and why. This code is dense and a few diagrams and inline comments might help. I'll just assume it's all good I guess if this is only used in testing.
___
*[pkg/util/tracing/tracer_span.go, line 391 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mKFdG6KXSwDRJmrtW:-M3mKFdG6KXSwDRJmrtX:b7ccvrm) ([raw file](https://github.com/cockroachdb/cockroach/blob/171b4e817ffa5e46f5b828242fdbd4cddd82f65f/pkg/util/tracing/tracer_span.go#L391)):*
> ```Go
> 	// Merge ownLogs with childSpans.
> 	mergedLogs := make([]traceLogData, 0, len(ownLogs))
> 	timeMax := time.Date(2100, 0, 0, 0, 0, 0, 0, time.UTC)
> ```

This has a gross feel to it. And it doesn't seem very optimistic. What can we do to make this 2200?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46596)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2648 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mLv0s241af0h4utHn:-M3mWbvgER_XGwjm-54B:b-ok0sgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c8fc6fb2795e21be9c7f99a93cf5a65f961fc3d/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2648)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why do we need the enum? Can't we just use a `roachpb.TransactionStatus`?
</blockquote></details>

because I only want a subset of the `TransactionStatus` here, and also I like reading the "expect" part.
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 2771 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mMw2J5CuMOdoyY2jB:-M3mXBEz7qST3ugjc5uN:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c8fc6fb2795e21be9c7f99a93cf5a65f961fc3d/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L2771)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> // The left side is CPut(a), the right side is CPut(b)+EndTxn(STAGING).
> ```

Consider moving this up to the `side` type definition.
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go, line 211 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mLdlo4pTAfemr12VQ:-M3mY3Db58mRt9Fjvv6U:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/00c32164199a475bbf96c2cf4f7cf0629c40064f/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go#L211)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Maybe we should include the error in here.
</blockquote></details>

done
___
*[pkg/util/tracing/tracer_span.go, line 264 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mIwWj4ymbPRy9JuKc:-M3mg-RN40OveJ_qfKMu:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/171b4e817ffa5e46f5b828242fdbd4cddd82f65f/pkg/util/tracing/tracer_span.go#L264)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Give this a comment.
</blockquote></details>

done
___
*[pkg/util/tracing/tracer_span.go, line 290 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mKddm4NNkcW8Ri4n4:-M3mmghk4l1mKf7vtwNK:bezzdgv) ([raw file](https://github.com/cockroachdb/cockroach/blob/171b4e817ffa5e46f5b828242fdbd4cddd82f65f/pkg/util/tracing/tracer_span.go#L290)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we write a test for this? Mostly I'm interested in seeing what effect it had.
</blockquote></details>

I'm gonna leave the test for next time :).
It looks like this. Before, messages from siblings were interspersed, sorted by their timestamp. Now children are inserted into parents, but otherwise spans are kept together.

```
     0.000ms      0.000ms    === operation:test trace
     0.002ms      0.002ms        === operation:dist sender send n:1
     0.009ms      0.007ms        event:[n1] querying next range at "b1"
     0.023ms      0.014ms        event:[n1] r33: sending batch 1 PushTxn to (n1,s1):1
     0.024ms      0.001ms        event:[n1] sending request to local client
     0.026ms      0.002ms            === operation:/cockroach.roachpb.Internal/Batch n:1
     0.028ms      0.002ms            event:[n1] node received request: 1 PushTxn
     0.035ms      0.007ms            event:[n1,s1] executing PushTxn(00000000->48691554) ["b1",/Min)
     0.039ms      0.004ms            event:[n1,s1,r33/1:{b-/Table/Syste…}] read-write path
     0.044ms      0.005ms            event:[n1,s1,r33/1:{b-/Table/Syste…}] sequencing request
     0.046ms      0.002ms            event:[n1,s1,r33/1:{b-/Table/Syste…}] acquiring latches
     0.050ms      0.004ms            event:[n1,s1,r33/1:{b-/Table/Syste…}] applied timestamp cache
     0.101ms      0.051ms            event:[n1,s1,r33/1:{b-/Table/Syste…}] found txn in indeterminate STAGING state meta={id=48691554 key="b1" pri=0.02894318 epo=0 ts=1585694049.825806000,0 min=1585694049.825806000,0 seq=4} lock=true stat=STAGING rts=0,0 wto=false max=0,0 int=1 ifw=2
     0.120ms      0.019ms            event:[n1,s1,r33/1:{b-/Table/Syste…}] evaluated request
     0.125ms      0.005ms            event:[n1,s1,r33/1:{b-/Table/Syste…}] recovering txn 48691554 from indeterminate commit
     0.483ms      0.358ms            event:[n1,s1,r33/1:{b-/Table/Syste…}] recovered txn 48691554 with status: ABORTED
     0.486ms      0.003ms            event:[n1,s1,r33/1:{b-/Table/Syste…}] sequencing request
     0.487ms      0.001ms            event:[n1,s1,r33/1:{b-/Table/Syste…}] acquiring latches
     0.491ms      0.004ms            event:[n1,s1,r33/1:{b-/Table/Syste…}] applied timestamp cache
     0.513ms      0.022ms            event:[n1,s1,r33/1:{b-/Table/Syste…}] evaluated request
```
___
*[pkg/util/tracing/tracer_span.go, line 302 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mJcNN8u68bTZELhn6:-M3mgGz4EeYk_m0TZezp:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/171b4e817ffa5e46f5b828242fdbd4cddd82f65f/pkg/util/tracing/tracer_span.go#L302)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`strings.Builder`
</blockquote></details>

done
___
*[pkg/util/tracing/tracer_span.go, line 334 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mKMyc3XNS_rmSCOCs:-M3mgyyl51511U08LjTG:b-ezcu4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/171b4e817ffa5e46f5b828242fdbd4cddd82f65f/pkg/util/tracing/tracer_span.go#L334)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Give this a comment. Also, to be honest, I'm having a really hard time understanding what changed in this commit and why. This code is dense and a few diagrams and inline comments might help. I'll just assume it's all good I guess if this is only used in testing.
</blockquote></details>

Added a comment

https://www.google.com/url?sa=i&url=https%3A%2F%2Ftenor.com%2Fsearch%2Fsaul-goodman-gifs&psig=AOvVaw1IHm2qW-B_L1IlJm7pwpWA&ust=1585779958074000&source=images&cd=vfe&ved=0CAIQjRxqFwoTCJDZ0-rgxegCFQAAAAAdAAAAABAH
___
*[pkg/util/tracing/tracer_span.go, line 391 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mKFdG6KXSwDRJmrtW:-M3mh2ui63Qpe9oUJRGI:b6r6c02) ([raw file](https://github.com/cockroachdb/cockroach/blob/171b4e817ffa5e46f5b828242fdbd4cddd82f65f/pkg/util/tracing/tracer_span.go#L391)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This has a gross feel to it. And it doesn't seem very optimistic. What can we do to make this 2200?
</blockquote></details>

changed to 2200 :P


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 6 files at r12, 1 of 1 files at r13, 4 of 4 files at r15.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-:-M3nCwmlB7z3IZ7btC9M:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/util/tracing/tracer_span.go, line 391 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-M3mKFdG6KXSwDRJmrtW:-M3nCZUU08AlvIe2BDvk:bkeotob) ([raw file](https://github.com/cockroachdb/cockroach/blob/171b4e817ffa5e46f5b828242fdbd4cddd82f65f/pkg/util/tracing/tracer_span.go#L391)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

changed to 2200 :P
</blockquote></details>

Now we're talking.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46596#-:-M3q6z0sCtrg3MLCeKWs:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)



<!-- Sent from Reviewable.io -->

andreimatei commented:
bors r+

On Wed, Apr 1, 2020 at 10:46 AM craig[bot] <notifications@github.com> wrote:

> Build failed
>
>    - GitHub CI (Cockroach)
>    <https://teamcity.cockroachdb.com/viewLog.html?buildId=1843304&buildTypeId=Cockroach_UnitTests>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/46596#issuecomment-607292880>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC4C4NPWZ7MBIHMKPXGCETRKNHULANCNFSM4LT2EFHQ>
> .
>
